### PR TITLE
Add support for room upgrades

### PR DESCRIPTION
**Pull request by Half-Shot**
This PR adds support for following room upgrade tombstones.

### CODE REVIEW

File: lib/BridgedRoom.js [hunk 1]
```diff
@@ -91,6 +91,18 @@ BridgedRoom.prototype.unlinkMatrixRoom = function(matrixRoomId) {
     ]);
 };
 
+BridgedRoom.prototype.swapRoomIds = function(oldRoomId, newRoomId) {
+    if (this._portalMatrixRoomId === oldRoomId) {
```
Comment (benparsons) [inline lib/BridgedRoom.js]: If this is ever false it would be Very Bad, right? We should log in an `else`

File: lib/BridgedRoom.js [hunk 2]
```diff
@@ -91,6 +91,18 @@ BridgedRoom.prototype.unlinkMatrixRoom = function(matrixRoomId) {
...
+        this.setPortalMatrixRoomId(newRoomId);
+    }
+
+    const index = this._linkedMatrixRoomIds.indexOf(oldRoomId);
+    if (index) {
+        this._linkedMatrixRoomIds.splice(index, 1);
+        this.linkMatrixRoom(newRoomId);
```
Comment (benparsons) [inline lib/BridgedRoom.js]: The relationship between `linkMatrixRoom()` and `setPortalMatrixRoomId()` is confusing - seems like they should be wrapped together and the common logic made clearer

File: lib/Main.js [hunk 3]
```diff
@@ -63,6 +61,28 @@ function Main(config, registration) {
         registration,
         roomStore,
         userStore,
+        roomUpgradeOpts: {
+            migrateEntry: (entry, newId) => {
```
Comment (benparsons) [inline lib/Main.js]: Where is this called? I can't see any references elsewhere

File: lib/BridgedRoom.js (hunk 1 above)
Comment (Half-Shot) [reply lib/BridgedRoom.js]: Well, it could also be a plumbed room.

File: lib/Main.js (hunk 3 above)
Comment (Half-Shot) [reply lib/Main.js]: https://github.com/matrix-org/matrix-appservice-bridge/blob/master/lib/components/room-upgrade-handler.js#L120

File: lib/BridgedRoom.js (hunk 2 above)
Comment (Half-Shot) [reply lib/BridgedRoom.js]: Yep, I believe one is a canonical roomId for portal rooms and the other one is for plumbed. Agreed these should not be run across different functions, but it feels like a problem I want to solve in another PR.

**Pull request merged**